client/DESIGN.md

```diff
@@ -0,0 +1,380 @@
+            { "actor": "Mark Hamill", "role": "Luke Skywalker" },
+            { "actor": "Carrie Fisher", "role": "Princess Leia Organa" },
+            { "actor": "Harrison Ford", "role": "Han Solo" } ] 
+    }
+
+We write the given JSON content to key "star-wars" in table "movie", provided that the current value associated with the "star-wars" key has remained unchanged since the time 1368151663681367 milliseconds.
+
+In addition, the DB server supports batch writes, and the cache will in prctice use batch writes instead of the traditional write format described above.
+
+There are several batch write formats the DB server accepts:
+
+##### JSON Array
+Simply wrap the multiple write requests in a JSON array, as shown below.
+
+    [ { "op": "create", "table": "movie", "key": "star-wars", value={ /* movie object */ } },
+      { "op": "delete", "table": "movie", "key": "stars-war" } } ]
+
+The rest of the HTTP request remains as described above.
+
+##### Request Pipelining
+
+Send each item in the batch write to the server as a separate HTTP request.  We indicate in the requests that they are part of a batch write with the item header as shown below in the example two-part batch write:
+
+    PUT /movie/star-wars HTTP/1.1
+    X-Item: 1/2
+
+    { /* movie object */ }
+
+    DELETE /movie/stars-war HTTP/1.1
+    X-Item: 2/2
+
+Unlike the JSON array batch write method, this Request Pipelining method will allow intermediate HTTP proxies to see that their cache entries for each item in the batch write are invalid.
+
+* **Question(1)**: I don't think Item is a real HTTP header (at least, not in Wikipedia...).  Can we just define a custom header as something like above?
+
+### Responses
+As with the Requests section, please see more information in [OMVCC and Caching](https://forum.treode.com/t/omvcc-and-http-caching/62).
+
+#### Read Respones
+
+The DB will respond to read requests with a message containing the desired value or an indication the value does not exist in the table. 
+
+In the success case, HTTP response for read should be similar to the following:
+
+    HTTP/1.1 200 Okay
+    Date: Wed, 14 Jan 2015 11:49:13 GMT
+    Last-Modified: Fri, 10 May 2013 02:07:43 GMT
+    Read-TxClock: 1421236153024853
+    Value-TxClock: 1368151663681367
+    Vary: Request-TxClock
+    {   "title": "Star Wars",
+        "cast": [
+            { "actor": "Mark Hamill", "role": "Luke Skywalker" },
+            { "actor": "Carrie Fisher", "role": "Princess Leia Organa" } ] }
+
+There are several cases in which the HTTP response does not include the JSON-formatted requested value.  Here are those cases:
+
+- **HTTP/1.1 304 Not Modified**: The requested value has not changed since the last request.  Note this response is still valuable to the cache, as it allows the cache to update the read time (R_t) for this entry to the read\_timestamp of this request.  (Between when the value was written and this read\_timestamp, we know the value has not changed.)
```

Comment:
To get 304, the GET needs to include a Condition-TxClock. On write, Condition-TxClock means, write if not changed since. On read, Condition-TxClock means read if changed since.